src/storage/src/source/timestamper.rs

```diff
@@ -0,0 +1,325 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Timestamper using persistent collection
+use std::collections::{HashMap, VecDeque};
+use std::task::{Context, Waker};
+
+use anyhow::Context as _;
+use futures::{FutureExt, StreamExt};
+use mz_dataflow_types::client::controller::storage::CollectionMetadata;
+use timely::progress::{Antichain, Timestamp as _};
+use timely::PartialOrder;
+
+use mz_dataflow_types::sources::MzOffset;
+use mz_expr::PartitionId;
+use mz_ore::now::NowFn;
+use mz_persist_client::read::{ListenEvent, ListenStream, ReadHandle};
+use mz_persist_client::write::WriteHandle;
+use mz_persist_client::Upper;
+use mz_repr::{Diff, Timestamp};
+
+pub struct CreateSourceTimestamper {
+    name: String,
+    read_progress: Antichain<Timestamp>,
+    write_upper: Antichain<Timestamp>,
+    persisted_timestamp_bindings: HashMap<PartitionId, VecDeque<(Timestamp, MzOffset)>>,
+    write_handle: WriteHandle<(), PartitionId, Timestamp, Diff>,
+    read_handle: ReadHandle<(), PartitionId, Timestamp, Diff>,
+    timestamp_bindings_listener: ListenStream<(), PartitionId, Timestamp, Diff>,
+    now: NowFn,
+    waker: Waker,
+}
+
+impl CreateSourceTimestamper {
+    pub async fn initialize(
```

Comment:
This should probably be called `new` since that's the convention for constructor functions.

I think this is missing an `as_of` argument which is picked by the storage controller (my PR had it wired through). We must respect whatever `as_of` frontier is requested because the upstream might have compacted offsets before that time.